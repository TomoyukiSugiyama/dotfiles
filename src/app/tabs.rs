use ratatui::buffer::Buffer;
use ratatui::{layout::Rect, style::Color, style::Style, text::Line, widgets::Widget};
use strum::{Display, EnumIter};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Display, EnumIter)]
pub(crate) enum SelectedTab {
    #[default]
    #[strum(to_string = "Dotfiles")]
    Dotfiles,
    #[strum(to_string = "Execute")]
    Execute,
}

impl SelectedTab {
    pub(crate) fn title(self) -> Line<'static> {
        let line = Line::from(format!("  {self}  "));
        line.style(Style::new().fg(Color::White).bg(Color::Black))
    }

    pub(crate) fn next(self) -> Self {
        match self {
            SelectedTab::Dotfiles => SelectedTab::Execute,
            SelectedTab::Execute => SelectedTab::Dotfiles,
        }
    }
    pub(crate) fn previous(self) -> Self {
        match self {
            SelectedTab::Dotfiles => SelectedTab::Execute,
            SelectedTab::Execute => SelectedTab::Dotfiles,
        }
    }
}

impl Widget for SelectedTab {
    fn render(self, area: Rect, buffer: &mut Buffer) {
        match self {
            SelectedTab::Dotfiles => "Dotfiles".render(area, buffer),
            SelectedTab::Execute => "Execute".render(area, buffer),
        }
    }
}
