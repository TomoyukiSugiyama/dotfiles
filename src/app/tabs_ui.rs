use super::tabs::SelectedTab;
use ratatui::buffer::Buffer;
use ratatui::layout::Rect;
use ratatui::style::Color;
use ratatui::style::Style;
use ratatui::text::Line;
use ratatui::widgets::Tabs;
use ratatui::widgets::Widget;
use strum::IntoEnumIterator;

impl SelectedTab {
    pub(crate) fn title(self) -> Line<'static> {
        let title = match self {
            SelectedTab::Dotfiles => "Dotfiles",
            SelectedTab::Workflow => "Workflow",
        };
        Line::from(format!("  {title}  "))
    }

    pub(crate) fn render_tabs(self, area: Rect, buffer: &mut Buffer) {
        let titles = SelectedTab::iter()
            .map(|tab| tab.title())
            .collect::<Vec<Line>>();
        Tabs::new(titles)
            .style(Style::new().fg(Color::White))
            .highlight_style(Style::new().fg(Color::Black).bg(Color::White))
            .select(self as usize)
            .render(area, buffer);
    }
}

impl Widget for SelectedTab {
    fn render(self, area: Rect, buffer: &mut Buffer) {
        self.render_tabs(area, buffer);
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use ratatui::{backend::TestBackend, Terminal};

    #[test]
    fn test_render_tabs_workflow_selected() {
        let tab = SelectedTab::Workflow;
        let mut terminal = Terminal::new(TestBackend::new(50, 3)).unwrap();
        let result = terminal.draw(|frame| frame.render_widget(tab, frame.area()));
        
        // Just verify rendering doesn't panic
        assert!(result.is_ok());
    }

    #[test]
    fn test_render_tabs_dotfiles_selected() {
        let tab = SelectedTab::Dotfiles;
        let mut terminal = Terminal::new(TestBackend::new(50, 3)).unwrap();
        let result = terminal.draw(|frame| frame.render_widget(tab, frame.area()));
        
        // Just verify rendering doesn't panic
        assert!(result.is_ok());
    }
}
