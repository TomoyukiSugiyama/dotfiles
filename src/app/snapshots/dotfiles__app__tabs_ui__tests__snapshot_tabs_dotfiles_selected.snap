---
source: src/app/tabs_ui.rs
assertion_line: 104
expression: rendered
---
   Dotfiles   â”‚   Workflow
